### DISCUSSION TITLE
Kustomization support for K8s Jobs with TTL

### DISCUSSION THREAD

**Discussion by Tigrex** [fluxcd/flux2#3708] [Q&A]
Hi Flux experts,

I face the same issue as discussed previously here on year ago (https://github.com/fluxcd/flux2/discussions/2689):
My k8s Job has a TTL, once it is removed, Flux will apply it again, and so on forever. 

Do we have a way now to reconcile such job only once?
If I wrap it with a helm chart, would Helm controller avoid recreating the same job?

Thank you!

**Comment by makkes** [comment]
As stated in https://github.com/fluxcd/flux2/discussions/2689#discussioncomment-3048445, TTLs aren't supported.

With Helm you have a little more leverage when using [chart hooks](https://helm.sh/docs/topics/charts_hooks/).

**Comment by stefanprodan** [comment]
Hey @Tigrex may I ask why the need to TTL? Flux deleted old jobs before each update.

**Reply by Tigrex** [reply]
Thank you for asking! I'm about to explain the root problem.
Our application is based on k8s operator and CRDs. The CRD is used to run some batch workflows, and the custom resource does not support patch operation. Kubectl apply will only patch the custom resource itself, the completed pod will not be updated or restarted.
The CRD has a flag similar to k8s job TTL, such that setting it to true will cause the pod and the custom resource to be deleted after the batch workflow runs to completion.

Now we are using Flux to deploy this CR. We want to make sure each deployment is done only once, and git commits can push through new deployments properly. I'm thinking of a few possible solutions:
1. do not use the "ttl flag", each time create a new name for this custom resource. For this approach we need to find a way to make sure new deployment kicks in only when the previous one is completed properly. maybe use healthChecks?
2. use the "ttl flag", since new deployment is automatically ignored if the old one does not run to completion. however, now we face this reconciling multiple times issue and wonder if wrap it in a helm chart would help
3. for k8s jobs, flux kustomization has a configuration force:true to delete the old job. i wonder if it's possible that we make the CRD also immutable, then can we get the benefits of this flag to automatically delete the old CR

**Reply by stefanprodan** [reply]
>  i wonder if it's possible that we make the CRD also immutable, then can we get the benefits of this flag to automatically delete the old CR

I think this is worth trying, you can have a webhook that rejects the PATCH operations and returns an error that fits https://github.com/fluxcd/pkg/blob/main/ssa/utils.go#L256

**Comment by Tigrex** [comment]
Before I try the immutable CRD approach, wrapping the job inside a helm chart seems to work.
A completed job (or CR) will not be deployed again by helm.
